modernsyntax/classical_looping.cpp

```diff
@@ -28,7 +28,15 @@ void indexLoop(std::vector<Tool>& tools){
 }
 
 void iteratorLoop(std::vector<Tool>& tools){
-  for(std::vector<Tool>::const_iterator tool = tools.begin(), end = tools.end(); tool != end; ++tool){
+  for(auto tool = tools.begin(), end = tools.end(); tool != end; ++tool){
```

Comment:
`auto tool` creates an unecessary copy of each object. Did you consider references?
In addition the `const_iterator` is const, while `auto` alone does not preserve constness.

Reply:
What exactly does `auto tool` copy? The input vector itself is already passed by reference.

How can `tool` be `const` when it is incremented? Supposedly `const_iterator` implies the _target_ to be `const`, which is not consistent with the declaration of the function input.

Reply:
What you get in hand is the object, not an iterator to it. `auto tool` copies every single tool you are looping over. `auto& tool` would prevent that.
And as you are dealing with the object and not an iterator, it can as well be const.

Reply:
The iterator increment `++tool` does not work with `const`, `auto` seems to favour the normal iterator structure over `const_iterator`.

Reply:
ah. sorry. you are right. I was having a different example in my head :-)